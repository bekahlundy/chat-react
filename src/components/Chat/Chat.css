.Chat--container {
  border: 1px solid #eaeaea;
  box-shadow: 1px 1px 12px lightgrey;
  display: flex;
  flex-direction: column;
  height: 500px;
  padding: 30px;
  width: 70%;
}

.Chat--header {
  font-size: 42px;
  padding-bottom: 20px;
  text-align: center;
  width: 100%;
}

.Chat--body {
  border: 1px solid black;
  display: flex;
  height: 100%;
}

.Chat--current-users {
  border: 1px solid black;
  padding: 10px;
  width: 20%;
}

.Chat--messages {
  height: 420px;
  position: relative;
  width: 80%;
}
