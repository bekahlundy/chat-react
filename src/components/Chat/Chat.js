import React from "react";
import { Messages } from "../Messages/Messages";
import { EnterChat } from "../EnterChat/EnterChat";
import { AddMessage } from "../AddMessage/AddMessage";
import { CurrentUsers } from "../CurrentUsers/CurrentUsers";
import io from "socket.io-client";
import axios from "axios";
import "./Chat.css";

export class Chat extends React.Component {
  constructor() {
    super();
    this.state = {
      message: "",
      messages: [],
      user: "",
      users: []
    };

    this.socket = io("localhost:8080");

    this.socket.on("RECEIVE_MESSAGE", msgObj => {
      this.receiveMessage(msgObj);
    });

    this.socket.on("NEW_USER_JOINED", userObj => {
      this.newUserJoined(userObj);
    });

    this.socket.on("USER_DROPPED", updatedUserStore => {
      this.userDropped(updatedUserStore);
    });
  }

  handleBeforeUnload = () => {
    this.socket.emit("DROP_USER", {
      user: this.state.user
    });
  };

  componentWillMount() {
    window.addEventListener("beforeunload", this.handleBeforeUnload);
  }

  componentWillUnmount() {
    window.removeEventListener("beforeunload", this.handleBeforeUnload);
  }

  getInitialUsersInChat = () => {
    axios.get("/users").then(res => {
        this.setState({ users: res.data });
      })
      .catch(error => {
        console.log("error", error);
      });
  };

  getInitialMsgsInChat = () => {
    axios.get("/messages").then(res => {
      this.setState({ messages: res.data });
    });
  };

  userDropped = updatedUserStore => {
    this.setState({ users: updatedUserStore });
  };

  newUserJoined = userObj => {
    this.setState({ users: [...this.state.users, userObj] });
    this.getInitialUsersInChat();
    this.getInitialMsgsInChat();
  };

  receiveMessage = msgObj => {
    this.setState({ messages: [...this.state.messages, msgObj] });
  };

  postMessage = message => {
    axios.post("/add-new-message", {
        author: this.state.user,
        message
      })
      .then(res => {
        this.setState({ message: res.data.message });
      })
      .catch(error => {
        console.log("error", error);
      });
  };

  sendMessage = message => {
    this.socket.emit("SEND_MESSAGE", {
      author: this.state.user,
      message
    });
    this.postMessage(message);
    this.setState({ message: "" });
  };

  postUser = user => {
    axios.post("/add-new-user", {
        user
      })
      .then(res => {
        this.setState({ user: res.data.user });
      })
      .catch(error => {
        console.log("error:", error);
      });
  };

  enterChat = user => {
    this.socket.emit("ADD_NEW_USER", {
      user
    });
    this.postUser(user);
  };

  render() {
    const { messages, user, users } = this.state;

    if (!user) {
      return <EnterChat enterChat={this.enterChat} />;
    }

    return (
      <div className="Chat--container">
        <div className="Chat--header">Hello, {user}</div>
        <div className="Chat--body">
          <div className="Chat--current-users">
            <CurrentUsers currentUser={user} allUsers={users} />
          </div>
          <div className="Chat--messages">
            <Messages currentUser={user} messages={messages} />
            <AddMessage sendMessage={this.sendMessage} />
          </div>
        </div>
      </div>
    );
  }
}
