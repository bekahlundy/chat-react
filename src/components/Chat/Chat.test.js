import React from "react";
import ReactDOM from "react-dom";
import { Chat } from "./Chat";
import { render, cleanup } from "react-testing-library";
import uuid from "uuid";
import axiosMock from "axios";
import MockAdapter from 'axios-mock-adapter';

describe("Chat", () => {
  let axiosMock: MockAdapter;

  beforeEach(() => {
  axiosMock = new MockAdapter(axios);
  axiosMock.onPost(`/add-new-user`).reply(201, { name: 'Bekah', id: uuid.v4() });
});

  afterEach(() => {
    cleanup();
  });

  it("renders enter chat component when there is no user", () => {
    const { container } = render(<Chat />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders multiple components when there is a user", async () => {
    axiosMock.onGet(`/users`).reply(200, []);
    axiosMock.onGet(`/messages`).reply(200, []);

    const { container } = render(<Chat />);
    await enterChat(container);
    expect(container.firstChild).toMatchSnapshot();
  });
});
