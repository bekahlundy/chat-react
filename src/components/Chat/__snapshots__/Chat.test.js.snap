// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders enter chat component when there is no user 1`] = `
<div
  class="EnterChat--container"
>
  <div
    class="EnterChat--welcome-title"
  >
    Welcome to the chat. Enter your name to join
  </div>
  <div
    class="EnterChat--input-container"
  >
    <input
      class="EnterChat--input"
      data-testid="enter-chat-input"
      placeholder="Enter name here"
      type="text"
      value=""
    />
    <button
      class="EnterChat--btn"
      data-testid="enter-chat-button"
    >
      Enter Chat
    </button>
  </div>
</div>
`;
